### DISCUSSION TITLE
Address query fail on TREL

### DISCUSSION THREAD

**Discussion by fgervais** [openthread/openthread#9431] [Q&A]
I have a Thread network with 2 border routers using TREL as follow:

```
         +--------+
         |        |
         | Server |
         |        |
         +----+---+
              |
              |
              |
              |
              |
              |                                                          Ethernet Network
--------------+-------------+--------------------------------------+----------------------
                            |                                      |
                            |                                      |
                            |                                      |
                      +-----+-----+                          +-----+-----+
                      |           |                          |           |
                      |    BR1    |                          |    BR2    |
                      |           |                          |           |
                      +-----+-----+                          +-----+-----+
                            |                                      |
                            |                                      |
                            |                                      |
                    +-------+-------+                      +-------+-------+
                    |               |                      |               |
                    |  Thread Mesh  |                      |  Thread Mesh  |
                    |               |                      |               |
                    +---------------+                      +-------+-------+
                                                                   |
                                                                   |
                                                                   |
                                                               +---+----+
                                                               |        |
                                                               |  FTD1  |
                                                               |        |
                                                               +--------+
```

Note1: The `Server` has a static route to the mesh prefix (`fd04:2240:0:102::/64`) through `BR1`.

The scenario is:
1. `FTD1` sends a query to `Server`
2. Query reaches `Server`
3. `Server` answer through `BR1`
4. `BR1` doesn't know about the `FTD1` address so sends an address query to `ff03::2`
5. Not sure what happens with this request exactly
6. `BR1` times out address query
7. `BR1` drops frames for `FTD1`

Logs on `BR1`:
![image](https://github.com/openthread/openthread/assets/5411996/411dc99a-be17-430b-89ab-e887682ebd63)

Any idea where I should look to track down this issue?

**Comment by jwhui** [comment]
Are BR1 and BR2 connected via the same Thread topology? The output of the following commands should provide some more visibility:
- `neighbor table`
- `trel peers`

**Reply by fgervais** [reply]
`BR1`:

```
fgervais@nuc2:~$ docker exec -it `docker ps -f name=autostart-otbr --format '{{.ID}}'` ot-ctl neighbor table
| Role | RLOC16 | Age | Avg RSSI | Last RSSI |R|D|N| Extended MAC     | Version |
+------+--------+-----+----------+-----------+-+-+-+------------------+---------+
|   C  | 0x4001 |  83 |      -53 |       -55 |1|0|0| 2a34a509ab0fdc71 |       4 |
|   C  | 0x4027 | 207 |      -49 |       -48 |0|0|0| 3afd77aed911d6e6 |       4 |
|   C  | 0x403b |  56 |      -51 |       -52 |1|0|0| 9a41b0606d64ff8b |       4 |
|   R  | 0xf400 |  30 |      -20 |       -20 |1|1|1| 16b6d342aef84ca3 |       4 |

Done
fgervais@nuc2:~$ docker exec -it `docker ps -f name=autostart-otbr --format '{{.ID}}'` ot-ctl trel peers
| No  | Ext MAC Address  | Ext PAN Id       | IPv6 Socket Address                              |
+-----+------------------+------------------+--------------------------------------------------+
|   1 | 16b6d342aef84ca3 | dead00beef00cafe | [fd04:2240:0:1:dea6:32ff:fe52:6bbc]:60353        |
Done
fgervais@nuc2:~$ 
```

`BR2`:
```
fgervais@otbr2:~$ docker exec -it `docker ps -f name=autostart-otbr --format '{{.ID}}'` ot-ctl neighbor table
| Role | RLOC16 | Age | Avg RSSI | Last RSSI |R|D|N| Extended MAC     | Version |
+------+--------+-----+----------+-----------+-+-+-+------------------+---------+
|   C  | 0xf40a | 145 |      -31 |       -32 |1|0|0| 9aa7c440f364b357 |       4 |
|   R  | 0x4000 |   3 |      -20 |       -20 |1|1|1| 2efe7e806e8ad225 |       4 |

Done
fgervais@otbr2:~$ docker exec -it `docker ps -f name=autostart-otbr --format '{{.ID}}'` ot-ctl trel peers
| No  | Ext MAC Address  | Ext PAN Id       | IPv6 Socket Address                              |
+-----+------------------+------------------+--------------------------------------------------+
|   1 | 2efe7e806e8ad225 | 4224000000001000 | [fd04:2240:0:0:0:0:0:20]:40042                   |
Done
fgervais@otbr2:~$
```

**Reply by abtink** [reply]
They `trel peers` show that have different EXT PAN ID `dead00beef00cafe` and `4224000000001000`, so the two nodes are not considered to be part of same Thread network  and would not talk to each other over trel link.

Here are some things to investigate:
- Did the two nodes set up to be part of the same network (do they have same EXT PAN ID?)?
- If they have the same ext pan id, next question would be if/how the advertised trel service does not reflect it? Maybe something went wrong when publishing the trel service on infra-link mdns (e.g., TXT entries did not get correctly updated and/or old xpanid value is being cached/used?)

**Reply by fgervais** [reply]
Yes this is because, when I configure the network, I do the following:

1. I start BR1, it is in factory state, I set it up
2. I start BR2 which is also in factory state.
3. The first time BR1 sees BR2, it is in it's factory default state (hence the `dead00beef00cafe` EXT PAN ID in TREL peers info).
4. I setup BR2 the same as BR1.
5. The TREL peers info is stuck to the factory info for BR2. Although I'm not sure if it actually affect the network.

**Reply by fgervais** [reply]
You can see they actually have the same config:

BR1:
```
fgervais@nuc2:~$ docker exec -it `docker ps -f name=autostart-otbr --format '{{.ID}}'` ot-ctl dataset active
Active Timestamp: 1
Channel: 15
Channel Mask: 0x07fff800
Ext PAN ID: 4224000000001000
Mesh Local Prefix: fdb5:341:c736:7c8d::/64
Network Key: [...]
Network Name: home
PAN ID: 0x2240
PSKc: [...]
Security Policy: 672 onrc
Done
```

BR2:
```
fgervais@otbr2:~$ docker exec -it `docker ps -f name=autostart-otbr --format '{{.ID}}'` ot-ctl dataset active
Active Timestamp: 1
Channel: 15
Channel Mask: 0x07fff800
Ext PAN ID: 4224000000001000
Mesh Local Prefix: fdb5:341:c736:7c8d::/64
Network Key: [...]
Network Name: home
PAN ID: 0x2240
PSKc: [...]
Security Policy: 672 onrc
Done
```

**Reply by abtink** [reply]
> The TREL peers info is stuck to the factory info for BR2. Although I'm not sure if it actually affect the network.

This is the main issue. 

After BR2 changes it network (dataset) it should start advertising "trel service" on mDNS using the new info (as part of the TXT data of trel service). The question is why is this not happening? 
- I know OT code will request this using `otPlatTrelRegisterService()` platform API (which passes the new TXT records).
- This is triggered on xpan id changes
https://github.com/openthread/openthread/blob/5f8f05cab59a2dcaab5f7e655bb3a94cb9cb0207/src/core/radio/trel_interface.cpp#L122-L140

The other aspect can be that the other node does not correctly discover and/or report when TXT data gets changed for an existing mDNS service (somehow reports the old TXT data or ... ?). If this is the case, then it is a platform issue.

**Reply by fgervais** [reply]
Thank you I will have a look.
